add support for hint_override in mark_unbacked

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #162652

Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

bobrenjc93 committed:
add support for hint_override in mark_unbacked

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

This consolidates the approach to only use the hint override for size hints in contrast to the original approach which actually allowed guarding on the hint_override.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

This consolidates the approach to only use the hint override for size hints in contrast to the original approach which actually allowed guarding on the hint_override.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

This consolidates the approach to only use the hint override for size hints in contrast to the original approach which actually allowed guarding on the hint_override.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 commented on test/inductor/test_torchinductor.py:
```diff
@@ -10674,6 +10674,34 @@ def override(x):
+
+        @torch.compile
+        def override(x):
+            return x.sum(dim=0)
```
added test

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

This consolidates the approach to only use the hint override for size hints in contrast to the original approach which actually allowed guarding on the hint_override.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

This consolidates the approach to only use the hint override for size hints in contrast to the original approach which actually allowed guarding on the hint_override.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

ezyang reviewed (commented):
Help me understand why this prevents people from guarding/adding runtime asserts based on the hints. It seems unsafe to allow arbitrary queries to symbolic_hint to rely on this. You need to make some assertion at the use site that you promise to ONLY use this in a heuristic nature, and that your output is sound even if it varies the full range.

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

ezyang commented:
Thanks, and hope you are feeling better. My puzzlement now is, why is this flag not implied by the fallback argument at all these call sites?

bobrenjc93 commented:
> why is this flag not implied by the fallback argument at all these call sites?

My intuition is against muddying the waters since `fallback` is a global fallback (all unbacked symints get the same value) whereas the `use_user_provided_hint_override` flag opts into a unique user provided value on a per-unbacked basis. 

Are you implying deriving `use_user_provided_hint_override` via `use_user_provided_hint_override=fallback is not None`? I just worry that can be a confusing API since a user might pass in `fallback=0` and be surprised when autotuning happens with a different value. Admittedly this is still the case with `fallback=0 && use_user_provided_hint_override=True` but, at least in mind, it's much easier to reason through since the flags are separate.

Or perhaps I'm misunderstanding? If you have a cleaner API proposal, I'm all ears!

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 committed:
Update on "add support for hint_override in mark_unbacked"


Very similar to https://github.com/pytorch/pytorch/pull/161007 except now for mark_unbacked.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

bobrenjc93 commented:
Synced with @ezyang offline and collectively decided to consolidate on the `fallback` kwarg.

bobrenjc93 commented:
@pytorchbot merge
